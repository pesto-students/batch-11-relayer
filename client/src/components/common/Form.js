/* eslint-disable react/jsx-props-no-spreading */
import React, { createRef } from 'react';
import PropTypes from 'prop-types';
import {
  Form as ReactstrapForm,
  Input,
  Col,
  Button,
  FormGroup,
  Label,
} from 'reactstrap';

const Form = ({ onClick, action, formFields }) => {
  const refs = [];
  const onSubmit = (event) => {
    // event.preventDefault();
    onClick(event, ...refs);
  };

  const RenderFields = () => formFields.map((field, idx) => {
    const { button, label, input } = field;
    const ref = createRef();
    const key = idx.toString();
    if (field.inputType === 'input') {
      refs.push(ref);
      return (
        <FormGroup row key={key}>
          <Label htmlFor={label.id} sm={12} md={2}>
            {label.label}
          </Label>
          <Col sm={12} md={5}>
            <Input {...input} innerRef={ref} />
          </Col>
        </FormGroup>
      );
    }
    return (
      <Button key={key} className={button.className} color={button.color}>
        {action}
      </Button>
    );
  });
  return (
    <ReactstrapForm onSubmit={onSubmit}>
      <RenderFields />
    </ReactstrapForm>
  );
};

Form.propTypes = {
  formFields: PropTypes.array,
}.isRequired;

export default Form;
