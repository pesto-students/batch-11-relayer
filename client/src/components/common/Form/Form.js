import React, { createRef } from 'react';
import PropTypes from 'prop-types';
import {
  Form as ReactstrapForm, Input, Col, Button, FormGroup, Label,
} from 'reactstrap';

const Form = ({ action, formFields }) => {
  const refs = [];
  const onSubmit = (event) => {
    event.preventDefault();
    formFields[formFields.length - 1].onClick(event, ...refs);
  };

  const RenderFields = () => formFields.map((field, idx) => {
    const {
      type, label, id, placeholder: filler, color, className,
    } = field;
    const ref = createRef();
    const key = idx.toString();

    if (field.inputType === 'input') {
      refs.push(ref);
      return (
        <FormGroup row key={key}>
          <Label htmlFor={id} sm={12} md={2}>
            {label}
          </Label>
          <Col sm={12} md={5}>
            <Input type={type} name={id} id={id} placeholder={filler} innerRef={ref} required />
          </Col>
        </FormGroup>
      );
    }
    return (
      <Button key={key} className={className} color={color}>
        {action}
      </Button>
    );
  });
  return (
    <ReactstrapForm onSubmit={onSubmit}>
      <RenderFields />
    </ReactstrapForm>
  );
};

Form.propTypes = {
  formFields: PropTypes.array,
}.isRequired;

export default Form;
