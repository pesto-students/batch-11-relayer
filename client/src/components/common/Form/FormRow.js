import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import {
  Col, FormGroup, Input, Label,
} from 'reactstrap';
import UserContext from '../../../contexts/UserContext';

const FormRow = (props) => {
  const {
    type, label, id, placeholder,
  } = props;
  const { refs, ref } = useContext(UserContext);

  React.useEffect(() => {
    refs.push(ref);
  }, [ref, refs]);

  return (
    <FormGroup row>
      <Label for={id} sm={12} md={2}>
        {label}
      </Label>
      <Col sm={12} md={5}>
        <Input type={type} name={id} id={id} placeholder={placeholder} innerRef={ref} />
      </Col>
    </FormGroup>
  );
};

FormRow.propTypes = {
  type: PropTypes.string,
  id: PropTypes.string,
  label: PropTypes.string,
  placeholder: PropTypes.string,
  ref: PropTypes.string,
}.isRequired;

export default FormRow;
