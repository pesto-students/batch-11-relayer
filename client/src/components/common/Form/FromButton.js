import React from 'react';
import PropTypes from 'prop-types';
import { Col, Button } from 'reactstrap';

const FormButton = (props) => {
  const {
    className, color, text,
  } = props;
  return (
    <Col>
      <Button className={className} color={color}>
        {text}
      </Button>
    </Col>
  );
};

FormButton.propTypes = {
  className: PropTypes.string,
  color: PropTypes.string,
  text: PropTypes.string,
}.isRequired;

export default FormButton;
