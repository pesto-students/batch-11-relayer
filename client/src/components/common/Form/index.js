import Form from './Form';
import FormRow from './FormRow';
import FormButton from './FromButton';

export { Form, FormRow, FormButton };
