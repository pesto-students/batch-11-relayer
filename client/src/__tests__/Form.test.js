/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { shallow } from 'enzyme';
import Form from '../components/common/Form';

describe('<Form />', () => {
  const onClick = jest.fn();
  const props = {
    onClick,
    action: '',
    formFields: [
      {
        inputType: 'input',
        label: { label: 'Email', id: 'email' },
        input: {
          type: 'email', name: 'email', placeholder: 'username@example.com', required: 'required',
        },
      }, {
        inputType: 'input',
        label: { label: 'Password', id: 'pass' },
        input: {
          type: 'password', name: 'pass', placeholder: 'P@ssw0rd', required: 'required',
        },
      }, {
        inputType: 'button',
        button: { color: '', className: '' },
      },
    ],
  };
  it('renders the Form component', () => {
    const wrapper = shallow(<Form {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
  test('calls onClick on Form submit', () => {
    const wrapper = shallow(<Form {...props} />);
    const form = wrapper.find('Form');
    form.simulate('submit');
    expect(onClick).toHaveBeenCalledTimes(1);
  });
});
