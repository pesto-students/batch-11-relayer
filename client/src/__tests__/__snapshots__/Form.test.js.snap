// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> renders the Form component 1`] = `ShallowWrapper {}`;
